erpnext/stock/doctype/stock_entry/stock_entry.py

```diff
@@ -803,6 +805,45 @@ def validate_batch(self):
 						if getdate(self.posting_date) > getdate(expiry_date):
 							frappe.throw(_("Batch {0} of Item {1} has expired.").format(item.batch_no, item.item_code))
 
+	def move_sample_to_retention_warehouse(self):
+		if self.purpose == "Material Receipt":
+			for item in self.get("items"):
+				if item.sample_quantity and item.batch_no:
+					retention_warehouse = frappe.db.get_single_value('Stock Settings', 'sample_retention_warehouse')
+					retainted_qty = get_batch_qty(item.batch_no, retention_warehouse, item.item_code)
+					max_retain_qty = frappe.get_value('Item', item.item_code, 'sample_quantity')
+					if retainted_qty >= max_retain_qty:
+						frappe.throw(_("Maximum Sample quantity - {0} have already been retained for Batch {1} and Item {2} in Batch {3}.").
+							format(retainted_qty, item.batch_no, item.item_code, item.batch_no), MaxSampleAlreadyRetainedError)
+					else:
+						sample_serial_nos = ''
+						if item.serial_no:
+							serial_nos = (item.serial_no).split()
+							if serial_nos and len(serial_nos) > item.sample_quantity:
+								serial_no_list = serial_nos[:-(len(serial_nos)-item.sample_quantity)]
+								sample_serial_nos = '\n'.join(serial_no_list)
+						create_sample_material_transfer(company = self.company, item = item, retention_warehouse = retention_warehouse, 
+							sample_serial_nos = sample_serial_nos)		
+
+def create_sample_material_transfer(company, item, retention_warehouse, sample_serial_nos):
+	stock_entry = frappe.new_doc("Stock Entry")
+	stock_entry.company = company
+	stock_entry.purpose = "Material Transfer"
+	stock_entry.append("items", {
+		"item_code": item.item_code,
+		"s_warehouse": item.t_warehouse,
+		"t_warehouse": retention_warehouse,
+		"qty": item.sample_quantity,
+		"basic_rate": item.basic_rate,
+		"conversion_factor": 1.0,
+		"serial_no": sample_serial_nos,
+		'batch_no': item.batch_no
+	})
+	stock_entry.insert()
+	stock_entry.submit()
```

Comment:
Creating and submitting stock_entry through code is never a good idea. Because, if users added some mandatory custom fields in Stock Entry, it will fail. And the user will never be able to submit the stock entry.

What we can do, instead of creating stock entry automatically, we can add a button "Make Retention Stock Entry". On click of button, it will prepare a new stock entry for all the items combined and load it.